src/Map.ts

```diff
@@ -0,0 +1,825 @@
+ *
+ * @since 1.14.0
+ */
+export const remove = <K>(S: Setoid<K>): (<A>(k: K, m: Map<K, A>) => Map<K, A>) => {
+  const lookupS = lookupWithKey(S)
+  return (k, m) => {
+    const found = lookupS(k, m)
+    if (found.isSome()) {
+      const r = new Map(m)
+      m.delete(k)
+      return r
+    }
+    return m
+  }
+}
+
+/**
+ * Delete a key and value from a map, returning the value as well as the subsequent map
+ *
+ * @since 1.14.0
+ */
+export const pop = <K>(S: Setoid<K>): (<A>(k: K, m: Map<K, A>) => Option<[A, Map<K, A>]>) => {
+  return (k, m) => {
+    const a = lookup(S)(k, m)
+    return a.isNone() ? none : some(tuple(a.value, remove(S)(k, m)))
+  }
+}
+
+/**
+ * Lookup the value for a key in a `Map`.
+ * If the result is a `Some`, the existing key is also returned.
+ * @since 1.14.0
+ */
+export const lookupWithKey = <K>(S: Setoid<K>): (<A>(k: K, m: Map<K, A>) => Option<[K, A]>) => {
+  return <A>(k: K, m: Map<K, A>) => {
+    const entries = m.entries()
+    let e: IteratorResult<[K, A]>
+    while (!(e = entries.next()).done) {
+      const [ka, a] = e.value
+      if (S.equals(ka, k)) {
+        return some(tuple(k, a))
+      }
+    }
+    return none
+  }
+}
+
+/**
+ * Lookup the value for a key in a `Map`.
+ * @since 1.14.0
+ */
+export const lookup = <K>(S: Setoid<K>): (<A>(k: K, m: Map<K, A>) => Option<A>) => {
+  const lookupWithKeyS = lookupWithKey(S)
+  return (k, m) => lookupWithKeyS(k, m).map(([_, a]) => a)
+}
+
+/**
+ * Test whether or not one Map contains all of the keys and values contained in another Map
+ *
+ * @since 1.14.0
+ */
+export const isSubmap = <K, A>(SK: Setoid<K>, SA: Setoid<A>): ((d1: Map<K, A>, d2: Map<K, A>) => boolean) => {
+  const lookupWithKeyS = lookupWithKey(SK)
+  return (d1: Map<K, A>, d2: Map<K, A>): boolean => {
+    const entries = d1.entries()
+    let e: IteratorResult<[K, A]>
+    while (!(e = entries.next()).done) {
+      const [k, a] = e.value
+      const d1OptA = lookupWithKeyS(k, d1)
+      const d2OptA = lookupWithKeyS(k, d2)
+      if (d1OptA.isNone() || d2OptA.isNone() || !SK.equals(k, d2OptA.value[0]) || !SA.equals(a, d2OptA.value[1])) {
+        return false
+      }
+    }
+    return true
+  }
+}
+
+/**
+ * @since 1.14.0
+ */
+export const empty = new Map<never, never>()
+
+/**
+ * @since 1.14.0
+ */
+export const getSetoid = <K, A>(SK: Setoid<K>, SA: Setoid<A>): Setoid<Map<K, A>> => {
+  const isSubmap_ = isSubmap(SK, SA)
+  return fromEquals((x, y) => isSubmap_(x, y) && isSubmap_(y, x))
+}
+
+/**
+ * Gets {@link Monoid} instance for Maps given {@link Semigroup} instance for their values
+ *
+ * @since 1.14.0
+ */
+export const getMonoid = <K, A>(SK: Setoid<K>, SA: Semigroup<A>): Monoid<Map<K, A>> => {
+  return {
+    concat: (mx, my) => {
+      const r = new Map(mx)
+      const entries = my.entries()
+      let e: IteratorResult<[K, A]>
+      while (!(e = entries.next()).done) {
+        const [k, a] = e.value
+        const lookupWithKeyS = lookupWithKey(SK)
```

Comment:
this can be moved outside the `return`

Reply:
I made this change. Marking it as resolved.